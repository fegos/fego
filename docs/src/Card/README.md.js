webpackJsonp([16],{462:function(a,t){a.exports={content:["article",["h1","\u5361\u7247 Card"],["h3","\u7ec4\u4ef6\u63cf\u8ff0"],["ul",["li",["p","\u4ee5\u5361\u7247\u7684\u5f62\u5f0f\u5c55\u793a\u4e00\u4e9b\u56fe\u6587\u6570\u636e\u3002"]]],["h3","\u793a\u4f8b\u4ee3\u7801"],["pre",{lang:"html",highlighted:'<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>Card</span> <span class="token attr-name">title</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>card title<span class="token punctuation">"</span></span> <span class="token attr-name">hoverable</span><span class="token attr-value"><span class="token punctuation">=</span>{false}</span><span class="token punctuation">></span></span>\n  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>p</span><span class="token punctuation">></span></span>card content<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>p</span><span class="token punctuation">></span></span>\n  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>p</span><span class="token punctuation">></span></span>card content<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>p</span><span class="token punctuation">></span></span>\n  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>p</span><span class="token punctuation">></span></span>card content<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>p</span><span class="token punctuation">></span></span>\n<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Card</span><span class="token punctuation">></span></span>'},["code",'<Card title="card title" hoverable={false}>\n  <p>card content</p>\n  <p>card content</p>\n  <p>card content</p>\n</Card>']],["h2","API"],["h3","props\u5217\u8868"],["table",["thead",["tr",["th","\u5c5e\u6027"],["th","\u8bf4\u660e"],["th","\u7c7b\u578b"],["th","\u9ed8\u8ba4\u503c"]]],["tbody",["tr",["td","className"],["td","\u6837\u5f0f\u7c7b\u540d\uff0c\u53ef\u7528\u4e8e\u8986\u76d6\u9ed8\u8ba4\u6837\u5f0f"],["td","string"],["td"]],["tr",["td","title"],["td","\u5361\u7247\u6807\u9898"],["td","string"],["td"]],["tr",["td","border"],["td","\u5361\u7247\u662f\u5426\u6709\u8fb9\u6846"],["td","boolean"],["td","true"]],["tr",["td","hoverable"],["td","\u9f20\u6807\u79fb\u8fc7\u65f6\u662f\u5426\u6d6e\u8d77"],["td","boolean"],["td","true"]],["tr",["td","cover"],["td","\u5c01\u9762\u56fe\uff0c\u53ef\u4ee5\u662f\u56fe\u7247\u7684\u5730\u5740\uff0c\u4e5f\u53ef\u4ee5\u662felement"],["td","react node"],["td"]],["tr",["td","footer"],["td","\u5361\u7247\u5e95\u90e8\uff0c\u53ef\u63a5\u53d7\u6570\u7ec4\u3001\u5b57\u7b26\u4e32\u3001element\u7c7b\u578b\u6570\u7ec4"],["td","react node"],["td"]]]]],meta:{title:"Card",subTitle:"\u5361\u7247",filename:"src/Card/README.md"},toc:["ul",["li",["a",{className:"bisheng-toc-h1",href:"#\u5361\u7247-Card",title:"\u5361\u7247 Card"},"\u5361\u7247 Card"]],["li",["a",{className:"bisheng-toc-h3",href:"#\u7ec4\u4ef6\u63cf\u8ff0",title:"\u7ec4\u4ef6\u63cf\u8ff0"},"\u7ec4\u4ef6\u63cf\u8ff0"]],["li",["a",{className:"bisheng-toc-h3",href:"#\u793a\u4f8b\u4ee3\u7801",title:"\u793a\u4f8b\u4ee3\u7801"},"\u793a\u4f8b\u4ee3\u7801"]],["li",["a",{className:"bisheng-toc-h2",href:"#API",title:"API"},"API"]],["li",["a",{className:"bisheng-toc-h3",href:"#props\u5217\u8868",title:"props\u5217\u8868"},"props\u5217\u8868"]]]}}});