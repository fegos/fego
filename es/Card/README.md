---
title: Card
subTitle: 卡片
---

# 卡片 Card

### 组件描述
- 以卡片的形式展示一些图文数据。

### 示例代码

```html
<Card title="card title" hoverable={false}>
  <p>card content</p>
  <p>card content</p>
  <p>card content</p>
</Card>
```

## API

### props列表

| 属性 | 说明 | 类型 | 默认值 |
|----|-----|------|------|
| className | 样式类名，可用于覆盖默认样式 | string |  |
| title | 卡片标题 | string |  |
| border | 卡片是否有边框 | boolean | true |
| hoverable | 鼠标移过时是否浮起 | boolean | true |
| cover | 封面图，可以是图片的地址，也可以是element | react node |  |
| footer | 卡片底部，可接受数组、字符串、element类型数组| react node |  |
