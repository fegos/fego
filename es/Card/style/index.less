@import "../../style/mixin/index";
@card-prefix: ~"@{prefix}-card";
@boxPaddingH: 24px;
@boxPaddingV: 16px;

.heading() {
  font-size: 16px;
  padding: @boxPaddingV 0;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  font-weight: 500;
}

.@{card-prefix} {
  background-color: @component-background;
  color: @text-color;
  &-header,
  &-body,
  &-footer {
    border-top: @border-width-base @border-style-base @border-color-split;
    &:first-child{
      border-top: none;
    }
  }
  &-border {
    border: @border-width-base @border-style-base @border-color-base;
    border-radius: @border-radius-base;
  }
  &-hoverable {
    transition: all @ease-in-out .3s;
    &:hover {
      box-shadow: @shadow-2;
    }
  }
  &-header {
    padding: 0 @boxPaddingH;
    vertical-align: middle;
    .@{card-prefix}-title {
      .heading();
    }
  }
  &-cover {
    & + .@{card-prefix}-body {
      border-top: none;
    }
    & > * {
      width: 100%;
      display: block;
    }
    & > img {
      vertical-align: middle;
      border-top-left-radius: @border-radius-base;
      border-top-right-radius: @border-radius-base;
    }
    &-no-radius {
      & > img {
        border-top-left-radius: 0;
        border-top-right-radius: 0;
      }
    }
  }
  &-body {
    padding: @boxPaddingV @boxPaddingH;
  }
  &-footer {
    vertical-align: middle;
    &-txt {
      .heading();
      padding: @boxPaddingV @boxPaddingH;
    }
    ul {
      display: flex;
      flex-direction: row;
      padding: @boxPaddingV 0;
      li {
        flex: 1;
        text-align: center;
        & + li {
          border-left: @border-width-base @border-style-base @border-color-split;
        }
      }
    }
  }
}
