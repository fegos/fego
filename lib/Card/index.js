'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Card = function (_React$Component) {
  _inherits(Card, _React$Component);

  function Card() {
    var _temp, _this, _ret;

    _classCallCheck(this, Card);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.state = {}, _temp), _possibleConstructorReturn(_this, _ret);
  }

  Card.prototype.renderHeader = function renderHeader() {
    var _props = this.props,
        prefixCls = _props.prefixCls,
        title = _props.title;

    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(prefixCls + '-header') },
      _react2.default.createElement(
        'div',
        { className: (0, _classnames2.default)(prefixCls + '-title') },
        title
      )
    );
  };

  Card.prototype.renderCover = function renderCover() {
    var _classNames;

    var _props2 = this.props,
        prefixCls = _props2.prefixCls,
        cover = _props2.cover,
        title = _props2.title;

    var imgEl = null;
    if (typeof cover === 'string') {
      imgEl = _react2.default.createElement('img', { src: cover, alt: '' });
    } else if (_react2.default.isValidElement(cover)) {
      imgEl = cover;
    }
    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(prefixCls + '-cover', (_classNames = {}, _classNames[prefixCls + '-cover-no-radius'] = title, _classNames))
      },
      imgEl
    );
  };

  Card.prototype.renderBody = function renderBody() {
    var _props3 = this.props,
        prefixCls = _props3.prefixCls,
        children = _props3.children;

    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(prefixCls + '-body') },
      children
    );
  };

  Card.prototype.renderFooter = function renderFooter() {
    var _props4 = this.props,
        prefixCls = _props4.prefixCls,
        footer = _props4.footer;

    var footEl = null;
    if (footer instanceof Array) {
      footEl = _react2.default.createElement(
        'ul',
        null,
        footer.map(function (item, i) {
          return _react2.default.createElement(
            'li',
            { key: item.id || item.key || i },
            item
          );
        })
      );
    } else if (typeof footer === 'string') {
      footEl = _react2.default.createElement(
        'div',
        { className: (0, _classnames2.default)(prefixCls + '-footer-txt') },
        footer
      );
    } else if (_react2.default.isValidElement(footer)) {
      footEl = footer;
    }
    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(prefixCls + '-footer') },
      footEl
    );
  };

  Card.prototype.render = function render() {
    var _classNames2;

    var _props5 = this.props,
        prefixCls = _props5.prefixCls,
        children = _props5.children,
        className = _props5.className,
        style = _props5.style,
        title = _props5.title,
        cover = _props5.cover,
        border = _props5.border,
        hoverable = _props5.hoverable,
        footer = _props5.footer;


    return _react2.default.createElement(
      'div',
      {
        className: (0, _classnames2.default)(prefixCls, className, (_classNames2 = {}, _classNames2[prefixCls + '-border'] = border, _classNames2[prefixCls + '-hoverable'] = hoverable, _classNames2)),
        style: style
      },
      title ? this.renderHeader() : null,
      cover ? this.renderCover() : null,
      children ? this.renderBody() : null,
      footer ? this.renderFooter() : null
    );
  };

  return Card;
}(_react2.default.Component);

Card.defaultProps = {
  prefixCls: 'ns-card',
  // 卡片标题
  title: '',
  // 是否有边框
  border: true,
  // 鼠标移过时是否浮起
  hoverable: true,
  // 封面图
  cover: '',
  // 底部
  footer: null
};
Card.propTypes = {
  prefixCls: _propTypes2.default.oneOf(['ns-card']),
  // 卡片标题
  title: _propTypes2.default.node,
  // 是否有边框
  border: _propTypes2.default.bool,
  // 鼠标移过时是否浮起
  hoverable: _propTypes2.default.bool,
  // 封面图
  cover: _propTypes2.default.node,
  // 底部，接收数组、字符串、节点三种类型
  footer: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.object), _propTypes2.default.string, _propTypes2.default.node])
};
exports.default = Card;