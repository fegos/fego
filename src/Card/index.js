import React from 'react';

export default class Card extends React.Component {
  render() {
    return <div>card</div>;
  }
}
