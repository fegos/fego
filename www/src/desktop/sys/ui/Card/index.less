.page {
  :global {
    .section {
      margin-right: 20px;
      width: 300px;
      float: left;
    }
  }
}
