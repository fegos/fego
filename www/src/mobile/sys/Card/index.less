.page {
  :global {
    .section {
      margin-bottom: 25px;
    }
  }
}
