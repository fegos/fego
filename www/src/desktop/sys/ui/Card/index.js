import React, { Component } from 'react';
import { Card, Icon } from 'fego';
import style from './index.less';

export default class Page extends Component {
  state = {}
  render() {
    return (
      <div className={style.page}>
        <div className="f-cb">
          <div className="section">
            <h3>简单卡片：</h3>
            <Card title="card title" hoverable={false}>
              <p>card content</p>
              <p>card content</p>
              <p>card content</p>
            </Card>
          </div>

          <div className="section">
            <h3>带底部文字的卡片：</h3>
            <Card title="card title" footer="card footer">
              <p>card content</p>
              <p>card content</p>
              <p>card content</p>
            </Card>
          </div>

          <div className="section">
            <h3>定制title和footer：</h3>
            <Card title={<Icon name="yes" />} footer={[<Icon name="plus" />, <Icon name="no" />]}>
              <p>card content</p>
              <p>card content</p>
              <p>card content</p>
            </Card>
          </div>

          <div className="section">
            <h3>带封面的卡片：</h3>
            <Card cover="https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1520502215780&di=f78971db8261dbf8c758bd6f40626d93&imgtype=0&src=http%3A%2F%2Fimg18.3lian.com%2Fd%2Ffile%2F201705%2F27%2F9ed0aa2c7f1f02128b64247254da5bec.jpg">
              <p>card content</p>
              <p>card content</p>
              <p>card content</p>
            </Card>
          </div>

          <div className="section">
            <h3>无边框的卡片：</h3>
            <Card title="card title" border={false} cover="https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1520502681319&di=5a1905ec9c7b6f9fb109afffa51b7abb&imgtype=0&src=http%3A%2F%2Fi0.hdslb.com%2Fbfs%2Farchive%2Ff4f4c038324150f08d916f26ba09d1277cb61b2a.jpg">
              <p>card content</p>
              <p>card content</p>
              <p>card content</p>
            </Card>
          </div>
        </div>
      </div>
    );
  }
}
