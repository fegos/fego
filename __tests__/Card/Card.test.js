import React from 'react';
import Renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Card from 'Card';

describe('Card test', () => {
  let ctn = null;
  const src = 'https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1521464375182&di=14fc6d89cec9170987aa044e6de24d15&imgtype=0&src=http%3A%2F%2Fp0.ifengimg.com%2Fhaina%2F2017_08%2Fd4713af347c14f5_w600_h375.jpg';
  beforeEach(() => {
    ctn = (
      <div>
        <p>card content</p>
        <p>card content</p>
        <p>card content</p>
      </div>
    );
  });

  afterEach(() => {
    ctn = null;
  });

  it('card snapshot', () => {
    const card = Renderer.create(<Card>{ctn}</Card>);
    expect(card.toJSON()).toMatchSnapshot();
  });

  it('card without border and hoverable', () => {
    const card = shallow(<Card border={false} hoverable={false}>{ctn}</Card>);
    expect(card.hasClass('ns-card-border')).toBe(false);
    expect(card.hasClass('ns-card-hoverable')).toBe(false);
  });

  it('card with string cover', () => {
    const card = shallow(<Card title="title" cover={src}>{ctn}</Card>);
    expect(card.find('img')).toHaveLength(1);
    expect(card.find('.ns-card-cover').hasClass('ns-card-cover-no-radius')).toBe(true);
  });

  it('card with element cover', () => {
    const card = shallow(<Card cover={<img src={src} alt="pic" />}>{ctn}</Card>);
    expect(card.find('img')).toHaveLength(1);
    expect(card.find('.ns-card-cover').hasClass('ns-card-cover-no-radius')).toBe(false);
  });

  it('card with array footer', () => {
    const card = shallow(<Card title="title" footer={['yes', 'no']}>{ctn}</Card>);
    expect(card.find('li')).toHaveLength(2);
    card.find('li').forEach((node, index) => {
      if (index === 0) {
        expect(node.text()).toEqual('yes');
      } else if (index === 1) {
        expect(node.text()).toEqual('no');
      }
    });
    // 直接这么写会报 TypeError: Cannot read property 'text' of undefined
    // 应该是取[0]/[1]之后节点的性质变了
    // 跟 $('div') 一样，取 $('div')[0] 之后变成了 js 的 dom 节点，就不能再应用 jquery 的方法了，如 $('div')[0].css();
    // expect(card.find('li')[0].text()).toEqual('yes');
    // expect(card.find('li')[1].text()).toEqual('no');
  });

  it('card with string footer', () => {
    const card = shallow(<Card title="title" footer="footer">{ctn}</Card>);
    expect(card.find('li')).toHaveLength(0);
    expect(card.find('.ns-card-footer-txt')).toHaveLength(1);
  });

  it('card with element footer', () => {
    const card = shallow(<Card title="title" footer={<span>footer</span>}>{ctn}</Card>);
    expect(card.find('li')).toHaveLength(0);
    expect(card.find('.ns-card-footer-txt')).toHaveLength(0);
  });
});
