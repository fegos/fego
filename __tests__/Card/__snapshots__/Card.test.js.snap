// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card test card snapshot 1`] = `
<div
  className="ns-card ns-card-border ns-card-hoverable"
  style={undefined}
>
  <div
    className="ns-card-body"
  >
    <div>
      <p>
        card content
      </p>
      <p>
        card content
      </p>
      <p>
        card content
      </p>
    </div>
  </div>
</div>
`;
